bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedComponentPropertiesReferenceProvider.java

```diff
@@ -0,0 +1,192 @@
+            }
+        }
+
+        return references;
+    }
+
+
+    /**
+     * Collect the resourceType of current resource and its children recursively.
+     * Avoided the usage of resource.getResourceType() as it returns the "jcr:primaryType"
+     * when the resource doesn't have "sling:resourceType" property.
+     *
+     * @param resource {@link Resource}
+     */
+    private Set<String> collectComponentResourceTypes(Resource resource) {
+        Set<String> componentResourceTypes = new HashSet<>();
+        if (resource != null) {
+            String componentResourceType = resource.getValueMap()
+                .get(JcrResourceConstants.SLING_RESOURCE_TYPE_PROPERTY, String.class);
+            if (StringUtils.isNotEmpty(componentResourceType)) {
+                componentResourceTypes.add(componentResourceType);
+            }
+
+            if (resource.hasChildren()) {
+                Iterator<Resource> children = resource.listChildren();
+                while (children.hasNext()) {
+                    componentResourceTypes.addAll(collectComponentResourceTypes(children.next()));
+                }
+            }
+        }
+        return componentResourceTypes;
+    }
+
+
+    /**
+     * Iterates through the given collection of resourceTypes and finds all the components that has
+     * shared dialog, then adds the root page of the given resource to the list.
+     *
+     * @param resource {@link Resource}
+     * @param componentResourceTypes {@link Set<String>}
+     * @return references
+     */
+    private List<Reference> findSharedComponents(Resource resource,
+        Set<String> componentResourceTypes) {
+        List<Reference> references = new ArrayList<>();
+
+        ResourceResolver resourceResolver = resource.getResourceResolver();
+        for (String resourceType : componentResourceTypes) {
+
+            Resource componentResource = getComponentResource(resourceType, resourceResolver);
+            if (componentResource == null) {
+                continue;
+            }
+
+            String jcrPrimaryType = componentResource.getValueMap().get(JcrConstants.JCR_PRIMARYTYPE, String.class);
+            if (StringUtils.isEmpty(jcrPrimaryType) || !jcrPrimaryType.equalsIgnoreCase("cq:Component")) {
+                continue;
+            }
+
+            Resource sharedDialog = componentResource.getResourceResolver()
+                .getResource(componentResource.getPath() + "/dialogshared");
+            Resource globalDialog = componentResource.getResourceResolver()
+                .getResource(componentResource.getPath() + "/dialogglobal");
+            if (sharedDialog == null && globalDialog == null) {
+                continue;
+            }
+
+            Page rootPage = pageRootProvider.getRootPage(resource);
+            if (rootPage != null) {
+                references.add(new Reference("sharedProperties", getReferenceName(rootPage),
```

Comment:
Before adding the reference, do we have to validate that the current user has the ability to publish the root page?  Or is that handled outside of this function by the general framework that identifies references?